Fixes for bottom layer

purdeaandrei opened the pull request:
1) Kicad mirrors the x axis for the bottom layer. This is reversed
2) When applying rotation adjustment to the bottom layer, it should be applied in the opposite direction

purdeaandrei commented:
Rebased to fix conflicts

matthewlai commented on jlc_kicad_tools/jlc_lib/cpl_fix_rotations.py:
```diff
@@ -66,11 +72,20 @@ def FixRotations(input_filename, output_filename, db):
+            rotation = (rotation + correction) % 360
+            rotation_changed = True
             break
+        if rotation_changed:
```
Why not always apply the change?

purdeaandrei replied:
I guess this if is not necessary for correct functionality, but:
1) It's more efficient. Doesn't have to run .format
2) It is the same behaviour from before this PR, where row[rotation_index] is modified only if it has to change. Before this PR there was no rotation_changed variable, but it was behaving the same because the row[rotation_index] = assignment was inside the above if statement. I didn't want to change unintended behaviour.

Removing this if would mean that say if rotation is not changed on a component, then the number format might still change.

matthewlai replied:
Cool thanks. I would say for simplicity let's just change it all the time. If it reveals a bug, it would have to be fixed anyways (since otherwise components on the back side would be buggy).

purdeaandrei replied:
Oh, I just looked at it again, and there may be a bug here,
looks like I forgot to do 'rotation_changed = True' on bottom components.
I'll investigate and update this PR.

purdeaandrei commented:
Please note, that I have successfully built a bottom-populated PCB with these changes:
https://deskthority.net/viewtopic.php?f=50&t=24512
I ended up not using resistor networks because they went out of stock for a while.

purdeaandrei commented:
![image](https://user-images.githubusercontent.com/544648/94371396-f2322d80-00fe-11eb-90b6-5ddedb60661c.png)
I did not get any placement errors from jlcpcb

purdeaandrei commented:
Alright, so the first version of this PR wasn't exactly right.
I misinterpreted what I was seeing, I thought I would need to add 180 degrees to rotation of bottom layer components, but that was wrong. It worked by coincidence for the project above, because I also had a bug in my code that only added 180 degrees to the components that would have been rotation-adjusted normally anyway. So and adjustment of +180 + 90 = +270 has the same effect as -90 would have, and that's why it looked correct.

So now the current version of the PR is correct, I tested it with many more components.
On the top layer it adds the adjustment, on the bottom layer it substracts the adjustment.

Also I removed that ugly rotation_changed variable. But still -- if rotation is not changed, the format of the rotation is not re-formatted. Do you want me to change that? It would involve just moving the "row[rotation_index] = "{0:.6f}".format(rotation)" statement out of the if and for blocks. I'm fine either way, but I think I would prefer to keep it like this.

(Note: also I removed a mod to the database from this PR, to make the purpose of the PR singular. I will put what I removed in a different PR. )

matthewlai commented on jlc_kicad_tools/jlc_lib/cpl_fix_rotations.py:
```diff
@@ -68,11 +74,19 @@ def FixRotations(input_filename, output_filename, db):
             _LOGGER.logger.info("Footprint {} matched {}. Applying {} deg correction"
                 .format(row[package_index], pattern.pattern, correction))
-            row[rotation_index] = "{0:.6f}".format((float(row[rotation_index]) + correction) % 360)
+            if side_index is not None and row[side_index].strip() == "bottom":
```
Can you add checks for posx_index and side_index up above instead (after the "if rotation_index is None" check)?

If we can't find any of these fields (likely because kicad changed the format) the output of the script will probably be wrong, and it would be good to die with an informative error message in that case, instead of silently ignoring it.

purdeaandrei replied:
Agreed, done!

matthewlai reviewed (changes requested):
Thanks for figuring this out!

matthewlai reviewed (commented):
Amazing thanks!
